bundles/io/org.eclipse.smarthome.io.json.gson/src/main/java/org/eclipse/smarthome/io/json/gson/internal/InternalGsonBindingService.java

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.io.json.gson.internal;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.smarthome.io.json.JsonBindingService;
+import org.eclipse.smarthome.io.json.gson.GsonBindingService;
+import org.eclipse.smarthome.io.json.gson.GsonTypeAdapterProvider;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * JSON data binding implementation that uses Google Gson to serialize and deserialize between
+ * Java objects and JSON.
+ *
+ * @author Flavio Costa - Initial contribution and API
+ *
+ * @param <T> Type to be serialized or deserialized.
+ */
+@Component(immediate = true, service = JsonBindingService.class, property = {
+        InternalGsonBindingService.FORMAT_OUTPUT_PROPERTY + ":Boolean=false" })
+public class InternalGsonBindingService<T> extends GsonBindingService<T> {
```

Comment:
The design of this class is mostly adapted from the existing [GsonProvider](https://github.com/eclipse/smarthome/blob/7176f8961a4caba501fe38bab52a44c8705011c3/bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/internal/GsonProvider.java) that is being replaced. The parameter `T` is not bound to anything because it is actually loaded by the classloader via the  `@Provider` annotation that JAX-WS relies on. If you check the interface `MessageBodyReader` you see the parameter `T` comes from there; we could bind it to Object, do you see any difference by doing so?

Right now, I don't see any other `GsonBindingService` implementation that would be done, but `InternalGsonBindingService` has an arbitrary way to set formatted (pretty) output and load `GsonTypeAdapterProviders`, so it is in a subclass... but there is really no strong reason to keep them separate.

Reply:
> The parameter `T` is not bound to anything because it is actually loaded by the classloader via the  `@Provider` annotation that JAX-WS relies on.

Can you explain this more detailed? I don't really understand.
Aren't generics only available on compile time because of the type erasure?

Reply:
Sure, but to put it briefly, I also don't think this will make any difference at runtime.

First of all, generics are not being used here because of a design choice in ESH, this parameter `T` comes from the interfaces MessageBodyReader and MessageBodyWriter that are implemented by our EntityProvider, so we have to do something with it: either replace it with a type specification, or leave undefined as it is the case now.

The parameter `T` is used in JAX-WS to constrain which types the EntityProvider supports; we could, for instance, have an entity provider for sitemaps and another one for things if we wanted. However, our code is absolutely generic so it works for any types served by the REST API. The expectation is that due to type erasure, at runtime the JVM will just take it as Object, which is fine as our code is generic for any object types. We could as well define the class as `EntityProvider<Object>` instead of  `EntityProvider<T>`; in theory it would be exactly the same at runtime, but to avoid introducing any unexpected bugs I just followed the way that the "ancestral" class was defined, namely `GsonProvider<T>`.